src/compiler/checker.ts

```diff
@@ -15036,6 +15036,26 @@ namespace ts {
             }
         }
 
+        // Return true if the given expression is possibly a discriminant value. We limit the kinds of
+        // expressions we check to those that don't depend on their contextual type in order not to cause
+        // recursive (and possibly infinite) invocations of getContextualType.
+        function isPossiblyDiscriminantValue(node: Expression): boolean {
+            switch (node.kind) {
+                case SyntaxKind.StringLiteral:
+                case SyntaxKind.NumericLiteral:
+                case SyntaxKind.NoSubstitutionTemplateLiteral:
+                case SyntaxKind.TrueKeyword:
+                case SyntaxKind.FalseKeyword:
+                case SyntaxKind.NullKeyword:
+                case SyntaxKind.Identifier:

```

Comment:
An identifier may end up being any type and cause a circularity still - while this change may happen to break the exact loop in the issue, there's _probably_ still a problem if you add an indirection.

Reply:
I'm not sure what you mean. The key here is that `checkIdentifier` will never ask for the contextual type of the identifier and therefore it can't cause a circularity issue.

Reply:
Also, literally any expression can be a discriminant here, see this, which works today and ths would break:
```ts
type A = {
    kind: "a",
    value: number
}

type B = {
    kind: "b",
    value: string
}

function getMode(): "a" {
    return "a";
}

const x: A | B = {
    kind: getMode(),
    value: 12
};

function needsNumber(x: number): void { }

needsNumber(x.value);
```